integrations/crewai/src/ucai_crewai/toolkit.py

```diff
@@ -0,0 +1,170 @@
+
+    @model_validator(mode="after")
+    def validate_toolkit(self) -> "UCFunctionToolkit":
+        client = validate_or_set_default_client(self.client)
+        self.client = client
+
+        if not self.function_names:
+            raise ValueError("Cannot create tool instances without function_names being provided.")
+
+        self.tools_dict = process_function_names(
+            function_names=self.function_names,
+            tools_dict=self.tools_dict,
+            client=client,
+            uc_function_to_tool_func=self.uc_function_to_crewai_tool,
+        )
+        return self
+
+    @staticmethod
+    def uc_function_to_crewai_tool(
+        *,
+        client: Optional[BaseFunctionClient] = None,
+        function_name: Optional[str] = None,
+        function_info: Optional[Any] = None,
+        **kwargs,
+    ) -> CrewAIBaseTool:
+        """
+        Converts a Unity Catalog function into a CrewAI tool.
+
+        Args:
+            client (Optional[BaseFunctionClient]): Client for executing the function.
+            function_name (Optional[str]): Name of the function to convert.
+            function_info (Optional[Any]): Detailed information of the function.
+
+        Note:
+            The following keys are popped from `kwargs` and used to instantiate 
+            the CrewAI BaseTool: ["description_updated", "cache_function", "result_as_answer"].
+            These values are not passed to the Unity Catalog function.
+
+        Returns:
+            CrewAIBaseTool: A tool representation of the Unity Catalog function.
+        """
+
+        if function_name and function_info:
+            raise ValueError("Only one of function_name or function_info should be provided.")
+        client = validate_or_set_default_client(client)
+
+        if function_name:
+            function_info = client.get_function(function_name)
+        elif function_info:
+            function_name = function_info.full_name
+        else:
+            raise ValueError("Either function_name or function_info should be provided.")
+            
+        crewai_kwargs_from_user = {}
+        for k in _CREWAI_KWARGS_FROM_USER:
+            if k in kwargs:
+                _logger.info(
+                    f"{k} is inferred to be a CrewAI BaseTool argument and will not be passed "
+                    "to the Unity Catalog function."
+                )
+                crewai_kwargs_from_user[k] = kwargs.pop(k)
+
+        def func(**kwargs: Any) -> str:
+            args_json = json.loads(json.dumps(kwargs, default=str))
+            result = client.execute_function(
+                function_name=function_name,
+                parameters=args_json,
+            )
+ 
+            return result.to_json()
+
+        generated_model = generate_function_input_params_schema(function_info).pydantic_model
+        class _BaseModelWrapper(generated_model):
+            #NB: Pydantic BaseModel.schema(), which is used by CrewAI, requires a subclass 
```

Comment:
fn_schema, the dataclass, has the property `pydantic_model` which inherits from `BaseClass`. Can we just pass `args_schema=generated_model.pydantic_model`?

Reply:
NB updated.